### ISSUE TITLE
Pagination throw 500 error with large page value

### ISSUE THREAD

**Issue by pbelyaev** [laravel/framework#17471]
- Laravel Version: 5.3.29
- PHP Version: 7.1.0
- Database Driver & Version: PostgreSQL 9.6.1

### Description:
If you would try to open page 999999999999999999 using paginate method it will throw 500 error:
``SQLSTATE[2201X]: <<Unknown error>>: 7 ERROR: OFFSET must not be negative``
It works the same for others website, e.g. laravel.io: https://laravel.io/forum/?page=999999999999999999

### Steps To Reproduce:
Create a route with the paginate method somewhere in it:
```php
Route::get('users', function () {
    return App\Users::paginate();
});
```

Then open the page 999999999999999999 and you'll get 500 error.

**Comment by hernandev** [comment]
TL;DR: You're seriously proposing that a page

```
nine hundred ninety nine quadrillion ,
nine hundred ninety nine trillion ,
nine hundred ninety nine billion ,
nine hundred ninety nine million ,
nine hundred ninety nine thousand ,
nine hundred ninety nine
```

this can't be realistic.

Correct me if I'm wrong, but an't database that would get close of storing that.
said that, i would recommend `won't fix` for this.

P.S.
On some logic of why that happens, PHP can't handle that number as integer, it will depend on platform and php version but PHP 7.1 on x64 limits integers to `PHP_INT_MAX` being `9223372036854775807`

This, **`9223372036854775807`** is kinda max number of records that will be parsed correctly on the offset.

Why? Numbers after that will be treat as float, that have a whole different representation, when too large (aka base 10).

so on those lines here

https://github.com/laravel/framework/blob/5.3/src/Illuminate/Database/Query/Grammars/Grammar.php#L579-L582

the framework is forcing the result as int, since that cannot be represented as int, it will reset on the max value and start on negative (really poor dumb explanation), that's why you got a negative number on the query.

There is a way of fixing it, but I do believe that the offset on MySQL also limits the max number to on offset approximately the same number.

So, give up or explain better why the hell you need this

**Comment by pbelyaev** [comment]
@hernandev Ok, it's obviously this could happen very rarely. Anyway, I don't like when anyone can throw 500 error on my website, especially If I have an alerts system that will wake me up at night.

**Comment by themsaid** [comment]
@pbelyaev what do you suggest?

**Comment by GrahamCampbell** [comment]
> especially If I have an alerts system that will wake me up at night.

Probably a bad idea to be alerted of every exception that occurs. Perhaps check out an error monitoring solution. I'm not going to suggest one here, but there are plenty that have laravel support.

**Comment by GrahamCampbell** [comment]
Throwing an exception when this happens is actually correct, however, the nature of this exception is undesirable, we should probably add another check and throw an exception before the db query fails. Alternatively, you could implement a middleware in your app to truncate page numbers on the request.

**Comment by hernandev** [comment]
@GrahamCampbell there are two easy solutions, but since Laravel does not cares on int size anywhere else (that I know), I would recommend just leaving it as it is.

## Solution 1:

Just make any offset bigger than max int, max int itself

```php
protected function compileOffset(Builder $query, $offset)
{
    // limit the max offset as being the max int size on the platform
    $offset = ($offset > PHP_INT_MAX) ? PHP_INT_MAX : $offset;

    return 'offset '.(int) $offset;
}
```

## Solution 2:

Throw a custom exception that will be handled nicely or ignored on the ExceptionHandler

```php
protected function compileOffset(Builder $query, $offset)
{
    if ($offset > PHP_INT_MAX) {
         throw new PaginationOutOfRangeOrWhatheverException("Page is out of max range") 
    }

    return 'offset '.(int) $offset;
}
````

**Comment by bartbrinkman** [comment]
While doing some negative testing some endpoints on my application I came across this too. For me the biggest issue is that `(int)` (or `intval()`) overflows on large numbers and can thus returns large negative numbers - which causes the code to throw database exceptions. Since the current solution is to 'just' cast it and deal silently with stuff like `?page=foo` (and/or actual SQL injection), I would suggest to just make sure its always positive _after_ casting.
